File: src/libres.domain/aggregates/Book.ts
import { Aggregate } from "../common/aggregate";
import { Error } from "../common/error";
import { Result } from "../common/result";
import { BookStatus } from "../enums/book-status";

export class Book extends Aggregate {
  private _name: string;
  private _description: string | null;
  private _price: number;
  private _order: number;
  private _bookStatus: BookStatus;
  private _accessLink: string;
  private _rating: number;
  private _categoryId: string | null;
  private _fingerPrint: string;
  private _userId: string;
  private _createdAt: Date;

  public get name(): string {
    return this._name;
  }
  public get description(): string | null {
    return this._description;
  }
  public get price(): number {
    return this._price;
  }
  public get order(): number {
    return this._order;
  }
  public get bookStatus(): BookStatus {
    return this._bookStatus;
  }
  public get accessLink(): string {
    return this._accessLink;
  }
  public get rating(): number {
    return this._rating;
  }
  public get categoryId(): string | null {
    return this._categoryId;
  }
  public get fingerPrint(): string {
    return this._fingerPrint;
  }
  public get userId(): string {
    return this._userId;
  }
  public get createAt(): Date {
    return this._createdAt;
  }

  protected constructor(
    id: string,
    name: string,
    accessLink: string,
    userId: string,
    categoryId: string | null,
    fingerPrint: string,
    description: string | null = "",
    price: number = 0,
    order: number = 0,
    bookStatus: BookStatus = BookStatus.PENDING,
    rating: number = 0,
  ) {
    super(id);
    this._name = name;
    this._accessLink = accessLink;
    this._userId = userId;
    this._categoryId = categoryId;
    this._fingerPrint = fingerPrint;
    this._description = description;
    this._price = price;
    this._order = order;
    this._bookStatus = bookStatus;
    this._rating = rating;
    this._createdAt = new Date();
  }

  public static reconstitute(
    id: string,
    name: string,
    accessLink: string,
    userId: string,
    categoryId: string | null,
    fingerPrint: string,
    description: string | null,
    price: number,
    order: number,
    bookStatus: BookStatus,
    rating: number,
  ): Book {
    return new Book(
      id,
      name,
      accessLink,
      userId,
      categoryId,
      fingerPrint,
      description,
      price,
      order,
      bookStatus,
      rating,
    );
  }

  public static create(
    name: string,
    accessLink: string,
    userId: string,
    categoryId: string | null,
    fingerPrint: string,
    description: string | null = "",
  ): Result<Book> {
    if (!name || name.trim() === "")
      return Result.Failure(Error.Validation("Name is required"));
    if (!accessLink || accessLink.trim() === "")
      return Result.Failure(Error.Validation("Access name is required"));

    const newId = crypto.randomUUID();

    return Result.Success(
      new Book(
        newId,
        name,
        accessLink,
        userId,
        categoryId,
        fingerPrint,
        description,
      ),
    );
  }

  public increaseOrder(): void {
    this._order++;
  }
  public reject(): void {
    this._bookStatus = BookStatus.REJECTED;
  }
  public accept(): void {
    this._bookStatus = BookStatus.ACCEPTED;
  }
}

File: apps/backend/src/Libres.API/Features/Books/Domain/Book.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Libres.API.Features.Users.Domain;
using Libres.API.Shared.Application.CustomError;
using Libres.API.Shared.Domain;
using Libres.API.Shared.Domain.Enums;

namespace Libres.API.Features.Books.Domain
{
    public class Book : AggregateRoot
    {
        public string Title { get; private set; } = string.Empty;
        public string? Description { get; private set; }
        public DateTime CreatedAt { get; private set; }
        // public User? Author { get; set; }
        public Guid UserId { get; set; }
        public decimal Price { get; set; } = 0;

        public int Order { get; private set; } = 0;
        public string? CoverImagePath { get; private set; }
        public string? FilePath { get; private set; }


        public Guid? CategoryId { get; private set; }

        public BookStatus BookStatus { get; private set; } = BookStatus.Pending;

        public double AverageRating { get; private set; } = 0.0;
        public int ReviewsCount { get; private set; } = 0;





        public Book() : base(Guid.Empty)
        {
        }

        public Book(Guid id, string title, Guid userId, Guid? categoryId,
                    decimal price,
                    string? description, string? coverImagePath, string? filePath) : base(id)
        {
            Title = title;
            Description = description;
            UserId = userId;
            CategoryId = categoryId;
            CreatedAt = DateTime.UtcNow;
            CoverImagePath = coverImagePath;
            Price = price;
            FilePath = filePath;
        }

        public static Result<Book> Create(string title, Guid userId, Guid? categoryId,
                                        decimal price,
                                          string? description, string? coverImagePath, string? filePath)
        {
            var book = new Book(Guid.NewGuid(), title, userId, categoryId, price, description, coverImagePath, filePath);
            return Result<Book>.Success(book);
        }


        public void UpdateStatus(BookStatus status)
        {
            this.BookStatus = status;
        }

        public Result ChangeName(string newName)
        {
            if (string.IsNullOrWhiteSpace(newName))
            {
                return Result.Failure("New Name Cannot be empty");
            }
            this.Title = newName;
            return Result.Success();
        }
        public Result ChangeDescription(string? newDescription)
        {
            this.Description = newDescription;

            return Result.Success();
        }
        public Result UpdatePrice(decimal newPrice)
        {
            if (newPrice < 0)
            {
                return Result.Failure($"{newPrice} < 0");
            }
            Price = newPrice;
            return Result.Success();
        }
        public Result ChangeCategory(Guid? newCategoryId)
        {
            CategoryId = newCategoryId;
            return Result.Success();
        }


        public Result<bool> UpdateAverageRate(double newAverageRating, int newReviewsCount)
        {
            if (newAverageRating < 0 || newAverageRating > 5)
            {
                return Result<bool>.Failure("Rating must be between 0 and 5.");
            }

            if (newReviewsCount < 0)
            {
                return Result<bool>.Failure("Reviews count cannot be negative.");
            }

            this.AverageRating = Math.Round(newAverageRating, 2);

            this.ReviewsCount = newReviewsCount;

            return Result<bool>.Success(true);
        }


    }
}